File: cs/src/ConstraintSolver/Intervals/ResetIntervals.cs
using System;
using System.IO;
using System.Threading;
using System.Collections.Generic;
//using Alica.Reasoner;
//using Al=Alica;
using AutoDiff;


namespace Alica.Reasoner.IntervalPropagation
{
		
	public class ResetIntervals : ITermVisitor<bool>
	{
		
		public ResetIntervals() {
		}
		
		
        public bool Visit(Constant constant) {			
			constant.Parents.Clear();
			UpdateInterval(constant,constant.Value,constant.Value);
			return true;
		}

        public bool Visit(Zero zero) {	
			zero.Parents.Clear();
			UpdateInterval(zero,0,0);
			return true;		
		}

        public bool Visit(ConstPower intPower) {
			intPower.Parents.Clear();
			
			intPower.Base.Accept(this);
			
			if (intPower.Exponent == 0) {
				UpdateInterval(intPower,0,0);
				return true;
			}
			double e = Math.Round(intPower.Exponent);
			if(intPower.Exponent == e && ((int)e)%2 == 0) {
				UpdateInterval(intPower,0,Double.PositiveInfinity);
				return true;
			}
			UpdateInterval(intPower,Double.NegativeInfinity,Double.PositiveInfinity);
			return false;
			
		}
		public bool Visit(TermPower tp) {
			tp.Parents.Clear();			
			tp.Base.Accept(this);
			tp.Exponent.Accept(this);
			UpdateInterval(tp,Double.NegativeInfinity,Double.PositiveInfinity);
			return false;
		}
		
		public bool Visit(Gp gp) {
			throw new NotImplementedException();
			return false;
		}
		
        public bool Visit(Product product) {
			product.Parents.Clear();
			UpdateInterval(product,Double.NegativeInfinity,Double.PositiveInfinity);
			product.Left.Accept(this);
			product.Right.Accept(this);			
			return false;
		}

        public bool Visit(Sigmoid sigmoid) {	
			sigmoid.Parents.Clear();
			sigmoid.Arg.Accept(this);
			sigmoid.Mid.Accept(this);			
			UpdateInterval(sigmoid,0,1);
			return true;
		}
		
		public bool Visit(LinSigmoid sigmoid) {	
			sigmoid.Parents.Clear();
			sigmoid.Arg.Accept(this);
			UpdateInterval(sigmoid,0,1);
			return true;
		}
		
        public bool Visit(LTConstraint constraint) {
			constraint.Parents.Clear();
			constraint.Left.Accept(this);
			constraint.Right.Accept(this);
			UpdateInterval(constraint,Double.NegativeInfinity,1);
			return true;
		}
		
        public bool Visit(LTEConstraint constraint) {
			constraint.Parents.Clear();
			constraint.Left.Accept(this);
			constraint.Right.Accept(this);
			UpdateInterval(constraint,Double.NegativeInfinity,1);
			return true;
		}
		
        public bool Visit(Min min) {
			min.Parents.Clear();
			min.Left.Accept(this);
			min.Right.Accept(this);			
			UpdateInterval(min,Double.NegativeInfinity,Double.PositiveInfinity);
			return true;
		}
		
		public bool Visit(Reification reif) {
			reif.Parents.Clear();
			reif.Condition.Accept(this);
			UpdateInterval(reif,reif.MinVal,reif.MaxVal);			
			return true;
		}
		
        public bool Visit(Max max) {
			max.Parents.Clear();
			max.Left.Accept(this);
			max.Right.Accept(this);			
			UpdateInterval(max,Double.NegativeInfinity,Double.PositiveInfinity);
			return true;
		}
		
        public bool Visit(And and) {
			and.Parents.Clear();
			and.Left.Accept(this);
			and.Right.Accept(this);
			UpdateInterval(and,Double.NegativeInfinity,1);
			//UpdateInterval(and,1,1); //enforce the purely conjunctive problem
			return true;
		}
		
        public bool Visit(Or or) {
			or.Parents.Clear();
			or.Left.Accept(this);
			or.Right.Accept(this);
			UpdateInterval(or,Double.NegativeInfinity,1);
			return true;
		}
		
        public bool Visit(ConstraintUtility cu) {
			cu.Parents.Clear();
			cu.Constraint.Accept(this);
			cu.Utility.Accept(this);			
			UpdateInterval(cu,1,Double.PositiveInfinity);
			return true;
		}
		
        public bool Visit(Sum sum) {
			sum.Parents.Clear();
			foreach(Term t in sum.Terms) t.Accept(this);
			UpdateInterval(sum,Double.NegativeInfinity,Double.PositiveInfinity);
			return true;
		}

        public bool Visit(AutoDiff.Variable variable) {
			variable.Parents.Clear();
			UpdateInterval(variable,variable.GlobalMin,variable.GlobalMax);
			return true;
		}

        public bool Visit(Log log) {
			log.Parents.Clear();
			log.Arg.Accept(this);
			UpdateInterval(log,Double.NegativeInfinity,Double.PositiveInfinity);
			return true;
		}
        public bool Visit(Sin sin) {
			sin.Parents.Clear();
			sin.Arg.Accept(this);
			UpdateInterval(sin,-1,1);
			return true;
		}

        public bool Visit(Cos cos) {
			cos.Parents.Clear();
			cos.Arg.Accept(this);
			UpdateInterval(cos,-1,1);
			return true;
		}

        public bool Visit(Abs abs) {
			abs.Parents.Clear();
			abs.Arg.Accept(this);
			UpdateInterval(abs,0,Double.PositiveInfinity);
			return true;
		}
		
 
        public bool Visit(Exp exp) {
			exp.Parents.Clear();
			exp.Arg.Accept(this);
			UpdateInterval(exp,0,Double.PositiveInfinity);
			return true;
		}
		public bool Visit(Atan2 atan2) {
			atan2.Parents.Clear();
			atan2.Left.Accept(this);
			atan2.Right.Accept(this);
			UpdateInterval(atan2,-Math.PI,Math.PI);			
			return true;
		}
		
		
		private void UpdateInterval(Term t, double min, double max) {			
			t.Min = min;
			t.Max = max;
			return;
		}
		
	}
	
}

File: java/src/main/java/Reasoner/IntervalPropagation/ResetIntervals.java
//
// Translated by CS2J (http://www.cs2j.com): 12.11.2016 09:40:47
//

package Reasoner.IntervalPropagation;
import AutoDiff.*;

//using Alica.Reasoner;
//using Al=Alica;
public class ResetIntervals implements ITermVisitor<Boolean>
{
    public ResetIntervals() throws Exception {
    }

    public Boolean visit(Constant constant) throws Exception {
        constant.Parents.clear();
        updateInterval(constant, constant.getValue(), constant.getValue());
        return true;
    }

    public Boolean visit(Zero zero) throws Exception {
        zero.Parents.clear();
        updateInterval(zero, 0, 0);
        return true;
    }

    public Boolean visit(ConstPower intPower) throws Exception {
        intPower.Parents.clear();
        intPower.getBase().accept(this);
        if (intPower.getExponent() == 0)
        {
            updateInterval(intPower, 0, 0);
            return true;
        }
         
        double e = Math.round(intPower.getExponent());
        if (intPower.getExponent() == e && ((int)e) % 2 == 0)
        {
            updateInterval(intPower, 0, Double.POSITIVE_INFINITY);
            return true;
        }
         
        updateInterval(intPower, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        return false;
    }

    public Boolean visit(TermPower tp) throws Exception {
        tp.Parents.clear();
        tp.getBase().accept(this);
        tp.getExponent().accept(this);
        updateInterval(tp, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        return false;
    }

    public Boolean visit(Product product) throws Exception {
        product.Parents.clear();
        updateInterval(product, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        product.getLeft().accept(this);
        product.getRight().accept(this);
        return false;
    }

    public Boolean visit(Sigmoid sigmoid) throws Exception {
        sigmoid.Parents.clear();
        sigmoid.getArg().accept(this);
        sigmoid.getMid().accept(this);
        updateInterval(sigmoid, 0, 1);
        return true;
    }

    public Boolean visit(LinSigmoid sigmoid) throws Exception {
        sigmoid.Parents.clear();
        sigmoid.getArg().accept(this);
        updateInterval(sigmoid, 0, 1);
        return true;
    }

    public Boolean visit(LTConstraint constraint) throws Exception {
        constraint.Parents.clear();
        constraint.getLeft().accept(this);
        constraint.getRight().accept(this);
        updateInterval(constraint, Double.NEGATIVE_INFINITY, 1);
        return true;
    }

    public Boolean visit(LTEConstraint constraint) throws Exception {
        constraint.Parents.clear();
        constraint.getLeft().accept(this);
        constraint.getRight().accept(this);
        updateInterval(constraint, Double.NEGATIVE_INFINITY, 1);
        return true;
    }

    public Boolean visit(Min min) throws Exception {
        min.Parents.clear();
        min.getLeft().accept(this);
        min.getRight().accept(this);
        updateInterval(min, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        return true;
    }

    public Boolean visit(Reification reif) throws Exception {
        reif.Parents.clear();
        reif.getCondition().accept(this);
        updateInterval(reif, reif.getMinVal(), reif.getMaxVal());
        return true;
    }

    @Override
    public Boolean visit(final Negation r) throws Exception
    {
        return true;
    }

    public Boolean visit(Max max) throws Exception {
        max.Parents.clear();
        max.getLeft().accept(this);
        max.getRight().accept(this);
        updateInterval(max, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        return true;
    }

    public Boolean visit(And and) throws Exception {
        and.Parents.clear();
        and.getLeft().accept(this);
        and.getRight().accept(this);
        updateInterval(and, Double.NEGATIVE_INFINITY, 1);
        return true;
    }

    //updateInterval(and,1,1); //enforce the purely conjunctive problem
    public Boolean visit(Or or) throws Exception {
        or.Parents.clear();
        or.getLeft().accept(this);
        or.getRight().accept(this);
        updateInterval(or, Double.NEGATIVE_INFINITY, 1);
        return true;
    }

    public Boolean visit(ConstraintUtility cu) throws Exception {
        cu.Parents.clear();
        cu.getConstraint().accept(this);
        cu.getUtility().accept(this);
        updateInterval(cu, 1, Double.POSITIVE_INFINITY);
        return true;
    }

    public Boolean visit(Sum sum) throws Exception {
        sum.Parents.clear();
        for (Object __dummyForeachVar0 : sum.getTerms())
        {
            Term t = (Term)__dummyForeachVar0;
            t.accept(this);
        }
        updateInterval(sum, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        return true;
    }

    public Boolean visit(AutoDiff.Variable variable) throws Exception {
        variable.Parents.clear();
        updateInterval(variable, variable.GlobalMin, variable.GlobalMax);
        return true;
    }

    public Boolean visit(Log log) throws Exception {
        log.Parents.clear();
        log.getArg().accept(this);
        updateInterval(log, Double.NEGATIVE_INFINITY, Double.POSITIVE_INFINITY);
        return true;
    }

    public Boolean visit(Sin sin) throws Exception {
        sin.Parents.clear();
        sin.getArg().accept(this);
        updateInterval(sin, -1, 1);
        return true;
    }

    public Boolean visit(Cos cos) throws Exception {
        cos.Parents.clear();
        cos.getArg().accept(this);
        updateInterval(cos, -1, 1);
        return true;
    }

    public Boolean visit(Abs abs) throws Exception {
        abs.Parents.clear();
        abs.getArg().accept(this);
        updateInterval(abs, 0, Double.POSITIVE_INFINITY);
        return true;
    }

    public Boolean visit(Exp exp) throws Exception {
        exp.Parents.clear();
        exp.getArg().accept(this);
        updateInterval(exp, 0, Double.POSITIVE_INFINITY);
        return true;
    }

    public Boolean visit(Atan2 atan2) throws Exception {
        atan2.Parents.clear();
        atan2.getLeft().accept(this);
        atan2.getRight().accept(this);
        updateInterval(atan2, -Math.PI, Math.PI);
        return true;
    }

    private void updateInterval(Term t, double min, double max) throws Exception {
        t.Min = min;
        t.Max = max;
        return ;
    }

}


